### PR TITLE
Remove Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll

### LINKED ISSUE #4149: Remove Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll from PowerShell core
This is a follow-up issue of #3649 
After refactoring PowerShell assembly load context, we actually don't need `Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll` anymore -- the remaining logic/functionality of this dll should be moved to `System.Management.Automation.dll` (`S.M.A.dll` depends on `AssemblyLoadContext.dll` today).
This might affect DSC because the DSC host is still depending on `AssemblyLoadContext.dll`, so we should communicate with DSC team on this to make sure nothing breaks.

Tasks
------

- [x] Remove `Microsoft.PowerShell.CoreCLR.AssemblyExtensions.LoadFrom(string assemblyPath)` once the PackageManagement module is migrated to .NET Core 2.0.
- [x] Removing `Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll` completely.

### PR DESCRIPTION
**Pull request by daxian-dbw**
Fix #4149

The code in `AssemblyLoadContext.dll` doesn't need to be in a separate DLL anymore.
S.M.A.dll depends on `AssemblyLoadContext.dll`, so keeping that code out of S.M.A.dll doesn't help make S.M.A smaller size or less dependent.
This PR is to move the code in `AssemblyLoadContext.dll` to `S.M.A.dll` and get rid of `AssemblyLoadContext.dll`.

The changes are:
- Move `CorePsAssemblyLoadContext.cs` to `src\S.M.A\CoreCLR\`
- Update `CorePsAssemblyLoadContext.cs` to get the test took moved to `Utils.InternalTestHooks` and update tests
- Update `build.psm1` and `.csproj` accrodingly
- Update `pwrshcommon.cpp` to remove `AssemblyLoadContext.dll` from the TPA list.
- `S.M.A.AssemblyExtensions` is removed as `PackageManagement` has finished their move to .NET Core 2.0. (I will work with Bryan to get the latest version uploaded to powershell-core)

Note that, the first commit is pure file movement. To easily see the changes to `CorePsAssemblyLoadContext.cs` after the move, you can just review the second commit.

### CODE REVIEW

File: src/System.Management.Automation/System.Management.Automation.csproj [hunk 1]
```diff
@@ -7,7 +7,6 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj" />
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/System.Management.Automation.csproj]: Need to remove the entry from Powershell-Win.sln as well.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/System.Management.Automation.csproj (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj]: Good point!

Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj]: Updated `PowerShell-Win.sln` and also deleted the file `Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj` (it was revived when I resolve conflicts).

**Review by adityapatwardhan — APPROVED** [review]
LGTM

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 2]
```diff
@@ -295,53 +295,52 @@ private Assembly Resolve(AssemblyLoadContext loadContext, AssemblyName assemblyN
...
         {
-            bool assemblyFound = false;
             assemblyFilePath = null;
+            if (Internal.InternalTestHooks.DisableGACLoading) { return false; }
+
+            bool assemblyFound = false;
             char dirSeparator = IO.Path.DirectorySeparatorChar;
 
-            if (PowerShellAssemblyLoadContextTestHooks.AllowGACLoading)
+            if(String.IsNullOrEmpty(_winDir))
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: looks like this file is inconsistent on whitespace after the `if`, at least for your changes, you should probably always have the whitespace

**Review by SteveL-MSFT — APPROVED** [review]

File: src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp [hunk 3]
```diff
@@ -1408,8 +1407,7 @@ namespace NativeMsh
         // Fall back to attempt to load the CLR from the alternate inbox location
         // or if the ALC was not located in the CoreCLR directory.
```
Comment (mirichmo) [inline src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp]: This comment should be updated as well

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 4]
```diff
@@ -295,53 +295,52 @@ private Assembly Resolve(AssemblyLoadContext loadContext, AssemblyName assemblyN
         // If there are multiple version of the assembly, load the latest.
         private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath)
         {
-            bool assemblyFound = false;
             assemblyFilePath = null;
+            if (Internal.InternalTestHooks.DisableGACLoading) { return false; }
```
Comment (mirichmo) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: Please spread it over 4 lines for consistency

```C#
if (Internal.InternalTestHooks.DisableGACLoading)
{
    return false;
}
```

**Review by mirichmo — CHANGES_REQUESTED** [review]

**Comment by mirichmo** [comment]
Just a few minor things. Otherwise it looks good.

File: src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp (hunk 3 above)
Comment (daxian-dbw) [reply src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp]: Sure thing. Also removed `std::string assemblyListToSearch = assemblyList.str();` in the next line as it's not used anywhere.

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs (hunk 4 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: Done.

**Review by mirichmo — APPROVED** [review]

**Comment by daxian-dbw** [comment]
An update of other offline works related to this change:
1. @brywang-msft has updated the `PackageManagement` module, and it works with this PR (feature CI passed).
2. I need to refresh the `psrp-windows` package. In progress.
3. I asked @nitinbps to validate this change with DSC native host. Haven't heard from him yet. I will check with him next Monday.

**Comment by adityapatwardhan** [comment]
@daxian-dbw Did you hear back from @nitinbps?

**Comment by daxian-dbw** [comment]
@adityapatwardhan Not yet, I will check with him today. I'm verifying the new `psrp.windows` packages at the meantime.

**Comment by daxian-dbw** [comment]
@adityapatwardhan talked to @nitinbps and he needs another 2 days to verify the changes.

**Comment by daxian-dbw** [comment]
Rebase to resolve the conflicts.

**Comment by daxian-dbw** [comment]
Updated and validated the [`psrp.windows` package](https://powershell.myget.org/feed/powershell-core/package/nuget/psrp.windows/6.0.0-beta.3). It works fine on Win10 and Win2012R2 (after install prerequisites). However, remoting doesn't work on Win7 as `Install-PowerShellRemoting.ps1` fails when running in powershell core on Win7:

image

I will sync with Mike tomorrow on this.
I also will sync with @nitinbps again tomorrow about DSC.

**Comment by mirichmo** [comment]
LGTM

**Pull request merged**